### PR TITLE
fix tensor detach to zero copy

### PR DESCRIPTION
**Pull request by zhouwei25**
### PR types

New features
### PR changes

APIs
### Describe

1. 修改了Tensor.detach，使得与原来Tensor共享数据内存；detach作为分离计算图的功能，可切断一部分计算图，仅将该Tensor数据进行前向计算，无需1次多余数据拷贝；
```
import paddle
import numpy as np

x = paddle.to_tensor(1.0, stop_gradient=False)
detach_x = x.detach()
y = x**2
y.backward()
print(x.grad)                 # [2.0]
print(detach_x.grad)    # None, stop_gradient=True

detach_x.stop_gradient = False   # Force stop_gradient to be False, supported auto-grad
z = detach_x**3
z.backward()
print(x.grad)               # [2.0], detach_x is detached from x's graph
print(detach_x.grad)  # [3.0]

detach_x[:] = 10.0
print(x.numpy())         # [10.0], detach_x share data with x
```

2. 与Tensor.clone的功能相关联，https://github.com/PaddlePaddle/Paddle/pull/27953

如果需要一个既分离计算图，又不共享数据内存的Tensor，请使用`x.detach().clone()`
```
import paddle
x = paddle.to_tensor(1.0, stop_gradient=False)
detach_clone_x = x.detach().clone()
```

### CODE REVIEW

File: paddle/fluid/pybind/imperative.cc [hunk 1]
```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
...
                 data = np.random.uniform(-1, 1, [30, 10, 32]).astype('float32')
-                with fluid.dygraph.guard():
-                    linear = Linear(32, 64)
-                    data = to_variable(data)
-                    x = linear(data)
-                    print(x.numpy())
+                linear = paddle.nn.Linear(32, 64)
+                data = paddle.to_tensor(data)
+                x = linear(data)
+                print(x.numpy())
```
Comment (liym27) [inline paddle/fluid/pybind/imperative.cc:661]: print(x.numpy()) -> print(x)

File: paddle/fluid/pybind/imperative.cc [hunk 2]
```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
...
-                                   "%s has not been initialized", self.Name()));
-             return self.NewVarBase(tensor.place(), false);
+           [](const imperative::VarBase
+                  &self) -> std::shared_ptr<imperative::VarBase> {
+             PADDLE_ENFORCE_EQ(
+                 self.Var().IsInitialized() &&
+                     (self.Var().IsType<framework::LoDTensor>() ||
+                      self.Var().IsType<framework::SelectedRows>()),
+                 true, platform::errors::InvalidArgument(
+                           "Tensor %s has not been initialized!", self.Name()));
```
Comment (liym27) [inline paddle/fluid/pybind/imperative.cc]: The message "Tensor %s has not been initialized!" doesn't match the requirement `self.Var().IsType<framework::LoDTensor>()  || self.Var().IsType<framework::SelectedRows>()`. Please split this `PADDLE_ENFORCE_EQ()` into two parts.

File: paddle/fluid/pybind/imperative.cc [hunk 3]
```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
...
+             VLOG(3) << "The detached Tensor(" << detach_var->Name()
+                     << ") share data with " << self.Name();
+             return detach_var;
            },
-           py::return_value_policy::copy, R"DOC(
+           py::return_value_policy::take_ownership, R"DOC(
 
         Returns a new Tensor, detached from the current graph.
+        It will share data with origin Tensor and always not have a Tensor copy and 
+        Tn addition, the detached Tensor does not provide gradient propagation.
```
Comment (liym27) [inline paddle/fluid/pybind/imperative.cc]: syntax error.

It will share data with origin Tensor and always not have a Tensor copy and Tn addition, the detached Tensor does not provide gradient propagation.

->

It will share data with origin Tensor and always **does** not have a Tensor copy. **In** addition, the detached Tensor does not provide gradient propagation.

File: paddle/fluid/pybind/imperative.cc [hunk 4]
```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
...
                 import paddle
-                linear = Linear(32, 64)
-                data = paddle.uniform(shape=[30, 10, 32], -1, 1)
-                x = linear(data)
-                y = x.detach()
+
+                x = paddle.to_tensor(1.0, stop_gradient=False)
+                detach_x = x.detach()
+                detach_x[:] = 10.0
+                print(x.numpy())      # [10.0], detach_x share data with x
```
Comment (liym27) [inline paddle/fluid/pybind/imperative.cc]: print(x.numpy()) -> print(x)

File: python/paddle/fluid/tests/unittests/test_var_base.py [hunk 5]
```diff
@@ -197,6 +197,33 @@ def test_tensor_to_variable(self):
...
+            self.assertEqual(detach_x.grad, None)
+
+            detach_x.stop_gradient = False  # Set stop_gradient to be False, supported auto-grad
+            z = 3 * detach_x**2
+            z.backward()
+            self.assertTrue(np.array_equal(x.grad, [20.0]))
+            print(x.grad)
+            print(detach_x.grad)
+            self.assertTrue(np.array_equal(detach_x.grad, [60.0]))
+            # Due to sharing of data with origin Tensor, There are some unsafe operations:
```
Comment (zhwesky2010) [inline python/paddle/fluid/tests/unittests/test_var_base.py]: This need you to remove annotation. @liym27

File: paddle/fluid/pybind/imperative.cc [hunk 6]
```diff
@@ -640,55 +640,109 @@ void BindImperative(py::module *m_ptr) {
...
+                z.backward()
+
+                print(x.grad)         # [20.0], detach_x is detached from x's graph, not affect each other
+                print(detach_x.grad)  # [300.0], detach_x has its own graph
+
+                # Due to sharing of data with origin Tensor， There are some unsafe operations:
+                y = 2 * x
+                detach_x[:] = 5.0
+                y.backward() 
+                # It will raise Error:
```
Comment (zhwesky2010) [inline paddle/fluid/pybind/imperative.cc:748]: Error message may be need to modified by you. @liym27

File: paddle/fluid/pybind/imperative.cc (hunk 2 above)
Comment (zhwesky2010) [reply paddle/fluid/pybind/imperative.cc]: Done

File: paddle/fluid/pybind/imperative.cc (hunk 3 above)
Comment (zhwesky2010) [reply paddle/fluid/pybind/imperative.cc]: Done

File: paddle/fluid/pybind/imperative.cc (hunk 4 above)
Comment (zhwesky2010) [reply paddle/fluid/pybind/imperative.cc]: Done

File: paddle/fluid/pybind/imperative.cc (hunk 1 above)
Comment (zhwesky2010) [reply paddle/fluid/pybind/imperative.cc:661]: this function is Tensor.numpy(), so should use Tensor.numpy()

File: paddle/fluid/pybind/imperative.cc (hunk 6 above)
Comment (liym27) [reply paddle/fluid/pybind/imperative.cc:748]: OK

**Review by liym27 — APPROVED** [review]
LGTM

**Review by TCChenlong — APPROVED** [review]
LGTM
